### DISCUSSION TITLE
Users - posts , likedPosts createdAt logs and notifications

### DISCUSSION THREAD

**Discussion by TamirCode** [prisma/prisma#19844] [Q&A]
I have many to many relations for the amount of likes a post has.
But I want to have data about when the like/dislike occured.
Should I give likes its own separate model? or is there a more elegant way?
Also what happens when a person takes back the like so its back to neutral

```prisma
model User {
    id        Int      @id @default(autoincrement())
    username  String   @unique
    password  String
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt

    // one to many
    posts    Post[]    @relation("posts")
    comments Comment[] @relation("comments")

    // many to many
    likedPosts       Post[]    @relation("likedPosts")
    dislikedPosts    Post[]    @relation("dislikedPosts")
    likedComments    Comment[] @relation("likedComments")
    dislikedComments Comment[] @relation("dislikedComments")
}

model Post {
    id        Int      @id @default(autoincrement())
    content   String
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt

    // many to one
    user     User      @relation("posts", fields: [userId], references: [id])
    userId   Int
    comments Comment[]

    // many to many
    likes    User[] @relation("likedPosts")
    dislikes User[] @relation("dislikedPosts")
}
```

**Comment by nurul3101** [comment]
Hello @TamirCode 👋 

If you want to store additional data about the likes/dislikes such as the timestamp when the like/dislike occurred, you should create an intermediate model for the many-to-many relationship. This is also called an "[explicit many-to-many relation](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations#explicit-many-to-many-relations)".

Your models could look something like this:
```prisma
model User {
  id        Int      @id @default(autoincrement())
  username  String   @unique
  password  String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // one to many
  posts Post[] @relation("posts")

  // explicit many to many
  postLikes    PostLike[]
  postDislikes PostDislike[]
}

model Post {
  id        Int      @id @default(autoincrement())
  content   String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // many to one
  user   User @relation("posts", fields: [userId], references: [id])
  userId Int

  // explicit many to many
  likes    PostLike[]
  dislikes PostDislike[]
}

model PostLike {
  id        Int      @id @default(autoincrement())
  userId    Int
  postId    Int
  createdAt DateTime @default(now())

  user User @relation(fields: [userId], references: [id])
  post Post @relation(fields: [postId], references: [id])
}

model PostDislike {
  id        Int      @id @default(autoincrement())
  userId    Int
  postId    Int
  createdAt DateTime @default(now())

  user User @relation(fields: [userId], references: [id])
  post Post @relation(fields: [postId], references: [id])
}
```
(Please note that I have removed comments related fields from these models for simplicity purposes, you can add them back if needed)

There are two models - `PostLike` and `PostDislike`, which act as the junction tables between `User` and `Post`. Each entry in `PostLike` or `PostDislike` represents a like or dislike by a user on a post and includes a timestamp (createdAt) for when the like or dislike occurred.

If a user takes back the like or dislike, you can delete the corresponding entry from the PostLike or PostDislike table.

Hope this helps! Please let me know in case you have any queries, if not, It would be great if you could **mark this Discussion as answered** to indicate that it has been resolved.

**Comment by TamirCode** [comment]
i will expand on my issue further, I want to have the data of when a dislike/like occurs and also when they are taken back. So I was thinking of creating a deletedAt column for each intermediate object instead of actually deleting it, and if the user likes again i guess it will create a new like with new id 
i am uncertain if this is an efficient way to log data, and on top of the logging data issue i also want users to have notifications of when a like happened, but unlike the logs, if a person takes away the like then it will truly delete from notification. 

title of discussion edited
```diff
- Users - posts , likedPosts createdAt
+ Users - posts , likedPosts createdAt logs and notifications
```

**Reply by nurul3101** [reply]
>  I want to have the data of when a dislike/like occurs and also when they are taken back. So I was thinking of creating a deletedAt column for each intermediate object instead of actually deleting it

This sounds like a good idea, you can create a column called `deletedAt` to soft delete the record instead of actually deleting it.

**Reply by nurul3101** [reply]
Do let us know in case you have any further queries.
